framework/src/meshgenerators/BlockToMeshConverterGenerator.C

```diff
@@ -0,0 +1,125 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "BlockToMeshConverterGenerator.h"
+#include "CastUniquePointer.h"
+#include "libmesh/elem.h"
+#include "MooseMeshUtils.h"
+
+registerMooseObject("MooseApp", BlockToMeshConverterGenerator);
+
+InputParameters
+BlockToMeshConverterGenerator::validParams()
+{
+  InputParameters params = MeshGenerator::validParams();
+
+  params.addClassDescription(
+      "Converts one or more blocks (subdomains) from a mesh into a stand-alone mesh with a "
+      "single block in it.");
+
+  params.addRequiredParam<MeshGeneratorName>("input", "The mesh we want to modify");
+  params.addRequiredParam<std::vector<SubdomainName>>(
+      "target_blocks",
+      "The (list of) blocks (or 'subdomains') we wish to have moved to a new mesh (by name, not "
+      "ID)");
+
+  return params;
+}
+
+BlockToMeshConverterGenerator::BlockToMeshConverterGenerator(const InputParameters & parameters)
+  : MeshGenerator(parameters),
+    _input(getMesh("input")),
+    _target_blocks(getParam<std::vector<SubdomainName>>("target_blocks"))
+{
+}
+
+std::unique_ptr<MeshBase>
+BlockToMeshConverterGenerator::generate()
+{
+  std::unique_ptr<MeshBase> mesh = std::move(_input);
+
+  if (!mesh->is_replicated())
+    mooseError("BlockToMeshConverterGenerator is not implemented for distributed meshes.");
+
+  auto new_mesh = buildMeshBaseObject();
+
+  std::vector<subdomain_id_type> target_block_ids =
+      MooseMeshUtils::getSubdomainIDs((*mesh), _target_blocks);
+
+  // Check that the block ids/names exist in the mesh
+  std::set<SubdomainID> mesh_blocks;
+  mesh->subdomain_ids(mesh_blocks);
+
+  for (std::size_t i = 0; i < target_block_ids.size(); ++i)
+    if (target_block_ids[i] == Moose::INVALID_BLOCK_ID || !mesh_blocks.count(target_block_ids[i]))
+    {
+      paramError("target_blocks",
+                 "The target_block '",
+                 getParam<std::vector<SubdomainName>>("target_blocks")[i],
+                 "' was not found within the mesh.");
+    }
+
+  // know which nodes have already been inserted, by tracking the old mesh's node's ids'
+  std::unordered_map<dof_id_type, dof_id_type> old_new_node_map;
+
+  for (subdomain_id_type target_block_id : target_block_ids)
+  {
+
+    for (auto elem : mesh->active_subdomain_elements_ptr_range(target_block_id))
```

Comment:
So if we have a uniformly refined mesh, this will flatten it, and if we have an adaptively refined mesh, this will break it?  The former behavior at least would need to be clearly documented, and the latter behavior is unacceptable; at the very least we need a test that would catch that and error out.

Reply:
I m not sure what you mean here. I just tried this
```
[Mesh]
  [file]
    type = FileMeshGenerator
    file = 3dmultiblock.e
  []
  [refine]
    type = RefineBlockGenerator
    input = file
    block = '2 4 6'
    refinement = '1 1 1'
  []
  [blockToMesh]
    type = BlockToMeshConverterGenerator
    input = refine
    target_blocks = "1 12 2 3 5 6"
  []
[]
```
and got the intended behavior afaik

Reply:
Let me try this input and see what I see.

Reply:
Okay, this time I don't think I'm mistaken; this is an actual bug.  Take a look at `n_elem()` and `n_active_elem()` in the console output.  I'm seeing 180 for both.  But in an adaptively refined mesh, those should never be the exact same number; any 2^dim level-1 active elements should have 1 non-active ancestor element, any 4^dim level-2 elements should have 2^dim+1 ancestors, etc.  Open the mesh here in paraview and it may look like an adaptively refined mesh, but libMesh will notice that those coarse elements don't have any neighbors on the side where the refined elements live, it'll happily conclude that it's looking at a slit mesh with internal boundaries at that interface, and at that point we can just pray the users don't believe their results.

Reply:
ok good catch!

I m just going to error out on these.

We dont currently run generators during a simulation, only at the very beginning.
Refinement at the beginning could just be done after the block extraction using this new MG so this is not that limiting either.

Reply:
Yeah, just exiting with an error is fine by me.  Having to reorder mesh generators is merely annoying; a mesh that looks correct but isn't would have been a disaster.